libsolidity/formal/Predicate.h

```diff
@@ -0,0 +1,124 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libsolidity/formal/SymbolicVariables.h>
+#include <libsolidity/formal/SymbolicVariables.h>
+
+#include <libsmtutil/Sorts.h>
+
+#include <map>
+#include <optional>
+#include <vector>
+
+namespace solidity::frontend
+{
+
+/**
+ * Represents a predicate used by the CHC engine.
+ */
+class Predicate
+{
+public:
+	static Predicate const* create(
+		smtutil::SortPointer _sort,
+		std::string _name,
+		smt::EncodingContext& _context,
+		ASTNode const* _node = nullptr
+	);
+
+	/// Predicate should be movable.
+	Predicate(Predicate&&) = default;
```

Comment:
If you have this, maybe you should also have ``Predicate& operator=(Predicate&&) = default;``? But fine without it, if things work without it...

The main thing I'm wondering is:
How can they be movable in the first place, if you have a static ``create`` function, which is supposed to keep them inside a static map :-)?
To me it seems that if you have that, then this cannot even be movable, since you don't want to ever move it out of this map...

Are they only movable, because they have to be move-constructible for the map ``emplace``?
If so, then just construct in-place instead, i.e. use ``m_predicates.emplace(std::piecewise_construct, std::forward_as_tuple(functorName), std::forward_as_tuple(_sort, _name, _context, _node));`` in the implementation of ``create`` instead - then you should be able to delete the move constructor (if not then there's something wrong somewhere, right?)

Reply:
Yea I wanted to add the move assign operator as well, but then it started complaining about `m_context` being a reference, so I left it out. Wouldn't be dangerous since the copy version is deleted anyway.

That does make sense.
I think in the first version, where the ownership was different, CHC also owned some, so I moved some predicates around.
But now I can't find any other move case. I'll try your suggestion.